### ISSUE TITLE
FreeBSD: use clang instead of g++ and other issues

### ISSUE THREAD

**Issue by NHOrus** [visualfc/liteide#523]
Starting from at least FreeBSD 10, compiler in base - clang.
So, correct spec goes go `-spec freebsd-clang`

Also, need to change qmake version depending on selected QT version - for QT4 it's `/usr/local/bin/qmake-qt4`, for QT5 it's `/usr/local/lib/qt5/bin/qmake`

To get valid build, I also need to invoke build script as LIBRARY_PATH=/usr/local/lib/ QTDIR=/usr/local/include/qt5 ./build_freebsd.sh

**Comment by wheelcomplex** [comment]
Hi @NHOrus , I have make a dirty hack to liteide and now it can build in freebsd 10(x64) with qt4+clang, would you please give it a try? here is it: https://github.com/wheelcomplex/liteide

**Comment by NHOrus** [comment]
Will try when get FreeBSD again.

**Comment by NHOrus** [comment]
For QT5 qmake lives in /usr/local/lib/qt5/bin/qmake

**Comment by NHOrus** [comment]
Works, but issue #524 persists. With QT5 build from ports. Shall try QT4 from packages... In a few hours.

**Comment by NHOrus** [comment]
QT4 from ports, segfault:

Starting program: /usr/home/nho/workshop/liteide/build/liteide/bin/liteide 
[New Thread 80c406400 (LWP 100392)]

Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 80c406400 (LWP 100392)]
0x00000008029726d6 in memcpy () from /lib/libc.so.7

**Comment by carlca** [comment]
Hello, I have just attempted to build under FreeBSD using the `wheelcomplex/liteide` build with `qt4` and `clang`. It seems to build but when I try to run the `liteide` executable using `./liteide`, I get the error message `Permission denied`. I build under su (root) and then did a `chown` and changed ownership to my local user. I also did a `chmod 755` on the entire build folder.
What am I doing wrong?

**Comment by wheelcomplex** [comment]
Sorry I didn’t follow this project for a long time.

**Comment by carlca** [comment]
No, it's alright. I've realised my mistake. I mistook the directory entry of `liteide` for an executable. I just need to sort out the `export LD_LIBRARY_PATH=$HOME/liteide/build/liteide/bin:$LD_LIBRARY_PATH` path setting and I'm sure it will work fine. Thanks for replying after all these years!

**Comment by dmgk** [comment]
@carlca Just FYI, LiteIDE is available as port and/or binary package on FreeBSD (devel/liteide, https://www.freshports.org/devel/liteide/)

**Comment by carlca** [comment]
Aha! Bless you Dmitri! I had been trying to build it and had got as far as an executable which ran, but where the code editing component was not visible, which rather ballsed things up!

I’ve just run `pkg install liteide` and it seems to have installed perfectly. I just need to find a way of installing some nice looking themes. Any ideas?

Thanks again, Dimitri, I was beginning to give up on getting liteide to work!

Cheers,
Carl

**Comment by carlca** [comment]
I spoke too soon. It looks like my env variables are messed up. My `$GOPATH` is set to `$HOME/go` while my `$GOROOT` is set to `/usr/local/go`. Yet when I go to the `Golang Playground` from the welcome page and click on Run, I get the following error...

![screen shot 2018-03-17 at 23 40 06](https://user-images.githubusercontent.com/260612/37561087-5e6d9f84-2a3d-11e8-91a0-ef304cfe282c.png)

Any ideas what the problem might be? Thanks!

**Comment by dmgk** [comment]
Hi Carl,

I'm not sure if LiteIDE supports loading themes from user home, but if it doesn't you can try copying theme .xml file into /usr/local/share/liteide/liteeditor/color and it should appear in Options/LiteApp/Theme dropdown.

I'm liteide FreeBSD port maintainer so if you encounter issues with running it on FreeBSD, please create a PR on https://bugs.freebsd.org or ping me directly in #freebsd on Freenode.

Good luck!
Dmitri

**Comment by dmgk** [comment]
@carlca If just you modified GOROOT/GOPATH in the shell env, you may need to relogin or try launching liteide from the terminal (looks like X session still has old env values).

**Comment by carlca** [comment]
I've rebooted my `Parallels` virtual machine and I've tried launching `LiteIDE` from the X desktop and from the terminal. In both cases I get the error shown in the screenshot above, despite the `$GOROOT` and `$GOROOT` being set correctly. Incidently the paths are set up the same in FreeBSD as they are in my host Apple Mac OS. LiteIDE runs perfectly including the Golang Playground in macOS.

**Comment by carlca** [comment]
Another problem is that if I open any old .go file and try to edit it, the editor behaves like the scroll lock key is switched on, on a Windows machine, and further to that, I cannot see the edit carat at all. Essentially I cannot edit any .go files with LiteIDE as it stands.

**Comment by dmgk** [comment]
@carlca Something must be set incorrectly because Go Playground integration works for me on FreeBSD 11.1-stable.

**Comment by carlca** [comment]
What are your $GOPATH and $GOROOT set to?

**Comment by carlca** [comment]
Does the fact that I am running FreeBSD as a guest OS under Parallels on an Apple Mac make any difference?

**Comment by dmgk** [comment]
It should not, my guess is that something in your Go env was not set up correctly. From the LiteIDE event log:

```
19:49:22 GolangPackage: GOROOT=/home/dmgk/opt/go
19:49:22 GolangPackage: GOPATH=/home/dmgk/go
```

**Comment by carlca** [comment]
This is strange. This is my event log from LiteIDE...

![screen shot 2018-03-18 at 00 57 23](https://user-images.githubusercontent.com/260612/37561479-7cddb6fc-2a47-11e8-9007-e54165675045.png)

Here is where I am setting the environment variables in my `$HOME/.profile` file...

![screen shot 2018-03-18 at 00 57 23](https://user-images.githubusercontent.com/260612/37561516-13f00dec-2a48-11e8-9223-f405201fef5c.png)

It seems that LiteIDE is not picking up the correct settings even though they are setup correctly (as far as I can tell) in `$HOME/.profile`

**Comment by dmgk** [comment]
What shell are you using (echo $SHELL)? Does env list GOROOT and GOPATH (env | grep ^GO)?
 ~/.profile is loaded by sh, if you're using csh, these settings should go to ~/.login.

**Comment by carlca** [comment]
$ echo $SHELL gives
/bin/sh

$ env | grep ^GO gives
GOROOT=/usr/local/go
GOPATH=/home/carlca/go

Notice as well in the screen dump of the event log, how near the top, the values of GOPATH and GOROOT are listed as /home/carlca/go

**Comment by carlca** [comment]
I'm gonna have to go to bed, as I am virtually falling asleep at my computer! Thanks for all your help so far. I'm sure the problem will be due to my lack of experience with FreeBSD (three days only!) but what the problem is I cannot imagine! Hopefully, chat with you in the morning.
Cheers,
Carl

**Comment by dmgk** [comment]
Oh ok, I see what's going on. Try selecting "system" from the environment combo in the toolbar.

**Comment by carlca** [comment]
Good morning Dimitri! Good news, I switched the environment combo to "system" and the playground works perfectly now. As for the other issue, it seems that LiteIDE thinks that the file is "read only"...

![screen shot 2018-03-18 at 12 49 34](https://user-images.githubusercontent.com/260612/37566015-42b3739a-2aab-11e8-8087-01e9a774bcac.png)

I've checked this by opening the file using edit making a change and saving it again. Also, a ls -l shows this...

![screen shot 2018-03-18 at 12 54 37](https://user-images.githubusercontent.com/260612/37566042-98c84e22-2aab-11e8-9b3d-f9f26633b4f2.png)

As far as I know that means that the file should be writable from as OS point of view. I'm still confused! I'm excited though because I'm sure the solution is quite simple and I'm looking forward to hacking in Golang with LiteIDE!

Thanks again!

Cheers,
Carl

**Comment by carlca** [comment]
Intriguingly, if I create a Golang file in LiteIDE, I am able to edit it just fine. Even more intriguingly, if I do an ls -l on that file it shows exactly the same attributes.

![screen shot 2018-03-18 at 14 46 26](https://user-images.githubusercontent.com/260612/37567193-372ea9e4-2abb-11e8-8941-f229a6273cab.png)

**Comment by carlca** [comment]
Aha! The first ls -l listing is showing root in the permissions column, so I guess this is a permissions issue. Just trying to suss out how to get chmod to do what I want...

**Comment by carlca** [comment]
I'm getting closer to sorting this. I tried running `chown carlcaulkett:carlcaulkett .` on the entire folder, but that didn't make any difference.

I then ran `chown carlcaulkett:carlcaulkett main.go` on the offending file itself. That worked. It seems that there is a problem in running `chmod` with wildcards.

**Comment by carlca** [comment]
Even putting into a bash script like this didn't work...

![screen shot 2018-03-18 at 22 32 06](https://user-images.githubusercontent.com/260612/37571990-4a83b7dc-2afc-11e8-8e3d-0a1f4819136f.png)

**Comment by carlca** [comment]
Sorted! `chown carlcaulkett:carlcaulkett ``ls``` did the trick!
